### ISSUE TITLE
Credentials expire when using Temporary STS Credentials

### ISSUE THREAD

**Issue by saskell** [stevearc/pypicloud#194]
Our setup here is that everyone uses Temporary STS Credentials that expire after a certain amount of time, eg in the ~/.aws/credentials file there is an aws_security_token and an expiration parameters in addition to the standard aws_access_key_id and aws_secret_access_key parameters.

I have a cron job that refreshes the credentials before they expire. 

I start the pypicloud server which finds these credentials and uses them with no problem...UNTIL the expiration time comes - at which point I start getting "Expired Token" errors.

I just simply restart the server and its good again.

I've run into this with other services - the only work around that I've come up with is to restart the service.  I was just wondering if you all had any better ideas on this issue.  

(And I'd be perfectly okay with you all saying "it's not our problem - it's a boto problem".)

Thanks!
Scott

**Comment by saskell** [comment]
Hi,
I waited 12 hours for my token to expire then I tried downloading a file through pypicloud's UI.

I thought maybe just maybe I'd see the exception in stdout/stderr but I saw nothing.

Is there an interface to generate a log file and/or change the log level?

Anyway, when I downloaded the file through pypicloud this is what I got in the browser:
```xml
<Error>
    <Code>ExpiredToken</Code>
    <Message>The provided token has expired.</Message>
    <Token-0>
        FQoGZXIvYXdzEH...
    </Token-0>
    <RequestId>C084...</RequestId>
    <HostId>
        n0fo...
    </HostId>
</Error>
```
Maybe I'll get a chance to dig into it more this weekend.

By the way - the package is pretty cool and very useful!

Thanks!
Scott

**Comment by stevearc** [comment]
Ohhhhhh...I see the problem now. It's not that it's throwing exceptions, it's that the generated URLs are expired. That makes this a bit harder.

The good news is that there is a [RefreshableCredentials](https://github.com/boto/botocore/blob/22f8beb1b177217bc436f82f83c471ea041161a0/botocore/credentials.py#L269) class in botocore that handles reloading the credentials from a source when they are near expiring.

The bad news is that it doesn't support your use case yet. `RefreshableCredentials` requires an [expiry_time](https://github.com/boto/botocore/blob/22f8beb1b177217bc436f82f83c471ea041161a0/botocore/credentials.py#L287-L289), but the shared credentials provider [doesn't parse Expiration from the file](https://github.com/boto/botocore/blob/22f8beb1b177217bc436f82f83c471ea041161a0/botocore/credentials.py#L1033-L1037)

I think that for this to work you'll need to make sure the botocore ingests the Expiration time. Without that we have no way of knowing if the credentials are still valid. And once it *can* ingest the expiration time, it should be easy (or easier) to put the auto-refresh logic into botocore itself.

**Comment by saskell** [comment]
Thanks for taking a look at it - the more I think about it, this should be handled by boto - trying to fix it in pypicloud would just "pollute" it.  

I checked the boto issue tracker and I did find a [feature request](https://github.com/boto/boto3/issues/1019) for adding a field to the credentials file to specify  an external process that refreshes credentials, after which boto would use the new credentials.  Unfortunately it's been open for the last 18 months.

In the mean time I'll write an external launcher that reads the expiration time, launches pypi cloud, kills it before the expiration time and if the credentials have been refreshed externally will re-launch pypi cloud.  A little brute force and kludgey but should work.

And I'll close this out.

Thanks!
Scott
